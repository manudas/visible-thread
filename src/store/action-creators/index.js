import {
    fetchTeamsScans,
} from '../../services';

export const FETCH_TEAMS_SCANS_TYPE = 'FETCH_TEAMS_SCANS';
export const FETCH_TEAMS_SCANS = (periods, division) => {
    return async (dispatch) => {

        const teamsData = await fetchTeamsScans(periods, division);

        dispatch( {
            type: FETCH_TEAMS_SCANS_TYPE,
            payload: teamsData
        });
    }
};
