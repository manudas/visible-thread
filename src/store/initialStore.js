import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

import {
    composeWithDevTools,
} from 'redux-devtools-extension';

import rootReducer from './reducers';

// no need of Thunk middleware as is not really asynchronous data fetching
export default createStore(rootReducer, composeWithDevTools(applyMiddleware(thunk)));
