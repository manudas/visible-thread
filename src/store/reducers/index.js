import { combineReducers } from 'redux';

import {
    fetchTeamsReducer,
} from './teams';

/**
 * The root reducer of the app
 */
export default combineReducers({
    teams: fetchTeamsReducer,
});
