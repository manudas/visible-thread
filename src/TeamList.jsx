import React, {
    useEffect, useState,
} from 'react';

import {
    useDispatch,
} from 'react-redux';

import DebouncedInput from './components/DebouncedInput';

import Switch from './components/Switch';

import { FETCH_TEAMS_SCANS } from './store/action-creators'

import DataTable from './components/DataTable';

import './TeamList.css';
import Popup from './components/Popup';
import { addTeam } from './services';


const TeamList = () => {
    const dispatch = useDispatch();
    const [scanNumber, setScanNumber] = useState(0);
    const [newTeamName, setNewTeamName] = useState(null);
    const [message, setMessage] = useState(null);
    const [isChecked, setChecked] = useState(false);

    useEffect(() => {
        // Taking this approach to show how working with Redux is, considering
        // that this data could be shared by more than one component, even if
        // it's not the case here
        dispatch(FETCH_TEAMS_SCANS(Number(scanNumber), !isChecked ? 'months': 'weeks'));
    } ,[dispatch, scanNumber, isChecked]);


    const toggleChecked = () => setChecked(!isChecked);

    const createNewTeamName = async ( name ) => {
        if (!name) {
            setMessage('Please type a name');
        } else {
            const ok = await addTeam(name);
            if (ok.toLowerCase() === 'true') {
                setMessage('Team added successfuly');
            } else {
                setMessage('Team wasn\'t added as it actually exists one team with the same name');
            }
        }
    }
    return <>
        <DebouncedInput onChangeHandler={(number) => setScanNumber(number)} label="Scans to fetch" />
        <Switch checked={ isChecked } onChange={ toggleChecked } optionLabels={['Weeks', 'Months']} />
        <DataTable />
        <div>
            <DebouncedInput onChangeHandler={(team) => setNewTeamName(team)} label="Create new team" />
            <button
                className="btn btn-primary m-5"
                onClick={() => createNewTeamName(newTeamName)}
            >Click to create</button>
        </div>
        <Popup text={message} closePopup={() => setMessage(null)} />
    </>;
}

export default TeamList;
