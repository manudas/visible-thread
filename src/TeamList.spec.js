import {
    render,
    fireEvent,
    waitFor
} from '@testing-library/react';

import { Provider } from 'react-redux'

import TeamList from './TeamList';

import store from './store/initialStore';
import { act } from 'react-dom/test-utils';

const ReduxProvider = ({ children, reduxStore }) => (
    <Provider store={reduxStore}>{children}</Provider>
);

describe('Test suite for TeamList', () => {

    // jest.useFakeTimers();

    let renderer;

    const sleep = (ms) => {
        return new Promise(resolveFn => setTimeout(resolveFn, ms));
    }

    beforeEach(() => {
        renderer = render(
                <ReduxProvider reduxStore={store}>
                <TeamList />
            </ReduxProvider>
        );
    });

    afterEach(() => {
        jest.restoreAllMocks(); // clears mocked implementations
        jest.clearAllMocks();
    });


    it('should render and be in the document', () => {
        const { container } = renderer;
        expect(container).toMatchSnapshot();
    });

    it('should search for months and weeks scans', async () => {
        const endpointSpyFunc = jest.spyOn(window, 'fetch').mockImplementation(async () => {
            return {
                ok: true,
                json: async () => [],
            };
        });

        const {
            getAllByTestId,
            getByTestId,
        } = renderer;

        // we have two, one for the months / weeks and another to add a new team
        const input = getAllByTestId('debouncedInput');

        await act(async() => {
            fireEvent.change(
                input[0], // input to select the number of months / weeks
                {
                    target:
                        {
                            value: "12"
                        }
                }
            )
            // debounceTimeout is 500
            // let's let the function to be run
            await sleep(600);
        });

        await waitFor(() => {
            expect(endpointSpyFunc).toHaveBeenCalledTimes(1); // called for months
        });
        // expect(endpointSpyFunc).ToHaveBeenCalledTimes(1); // called for months
        expect(endpointSpyFunc).toHaveBeenCalledWith('http://stubber.test.visiblethread.com/scans/monthly/12');

        const _switch = getByTestId('toggle-switch');
        fireEvent.click(_switch);

        await waitFor(() => {
            expect(endpointSpyFunc).toHaveBeenCalledTimes(2); // 2nd call is for weeks
        });

        expect(endpointSpyFunc).toHaveBeenCalledWith('http://stubber.test.visiblethread.com/scans/weekly/12');
    });

    it('cannot add team if team exists', async () => {
        const endpointSpyFunc = jest.spyOn(window, 'fetch').mockImplementation(async () => {
            return {
                ok: true,
                text: async () => "false",
            };
        });

        const {
            container,
            getAllByTestId,
            getByTestId,
            getByText,
        } = renderer;

        // we have two, one for the months / weeks and another to add a new team
        const input = getAllByTestId('debouncedInput');

        await act(async() => {
            fireEvent.change(
                input[1], // input to type the new team name
                {
                    target:
                        {
                            value: "Team 1"
                        }
                }
            )
            // debounceTimeout is 500
            // let's let the function to be run
            await sleep(600);
        });

        const button = getByText(/Click to create/);

        fireEvent.click(button);

        await waitFor(() => {
            expect(getByTestId('popup__container')).toBeInTheDocument();
        });

        await waitFor(() => {
            expect(getByText(/Team wasn't added as it actually exists one team with the same name/)).toBeInTheDocument();
        });

        expect(endpointSpyFunc).toHaveBeenCalled();
        expect(container).toMatchSnapshot();
    });

    it('Add a team if does not exists', async () => {
        const endpointSpyFunc = jest.spyOn(window, 'fetch').mockImplementation(async () => {
            return {
                ok: true,
                text: async () => "true",
            };
        });

        const {
            container,
            getAllByTestId,
            getByTestId,
            getByText,
        } = renderer;

        // we have two, one for the months / weeks and another to add a new team
        const input = getAllByTestId('debouncedInput');

        await act(async() => {
            fireEvent.change(
                input[1], // input to type the new team name
                {
                    target:
                        {
                            value: "Team 100"
                        }
                }
            )
            // debounceTimeout is 500
            // let's let the function to be run
            await sleep(600);
        });

        const button = getByText(/Click to create/);

        fireEvent.click(button);

        await waitFor(() => {
            expect(getByTestId('popup__container')).toBeInTheDocument();
        });

        await waitFor(() => {
            expect(getByText(/Team added successfuly/)).toBeInTheDocument();
        });

        expect(endpointSpyFunc).toHaveBeenCalled();
        expect(container).toMatchSnapshot();
    });
});
