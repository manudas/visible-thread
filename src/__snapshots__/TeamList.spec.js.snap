// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test suite for TeamList Add a team if does not exists 1`] = `
<div>
  <label
    class="debounced-input__label"
    for="subreddit"
  >
    Scans to fetch
    <input
      class="debounced-input__input"
      data-testid="debouncedInput"
      id="subreddit"
      value=""
    />
  </label>
  <div
    class="toggle-switch m-5"
    data-testid="toggle-switch"
  >
    <input
      class="toggle-switch-checkbox"
      id="switch"
      name="switch"
      readonly=""
      type="checkbox"
    />
    <label
      class="toggle-switch-label"
      for="switch"
    >
      <span
        class="toggle-switch-inner"
        data-no="Months"
        data-yes="Weeks"
      />
      <span
        class="toggle-switch-switch"
      />
    </label>
  </div>
  <div>
    <label
      class="debounced-input__label"
      for="subreddit"
    >
      Create new team
      <input
        class="debounced-input__input"
        data-testid="debouncedInput"
        id="subreddit"
        value="Team 100"
      />
    </label>
    <button
      class="btn btn-primary m-5"
    >
      Click to create
    </button>
  </div>
  <div
    class="popup__container"
    data-testid="popup__container"
  >
    <div
      class="popup__content bg-light"
    >
      Team added successfuly
    </div>
  </div>
</div>
`;

exports[`Test suite for TeamList cannot add team if team exists 1`] = `
<div>
  <label
    class="debounced-input__label"
    for="subreddit"
  >
    Scans to fetch
    <input
      class="debounced-input__input"
      data-testid="debouncedInput"
      id="subreddit"
      value=""
    />
  </label>
  <div
    class="toggle-switch m-5"
    data-testid="toggle-switch"
  >
    <input
      class="toggle-switch-checkbox"
      id="switch"
      name="switch"
      readonly=""
      type="checkbox"
    />
    <label
      class="toggle-switch-label"
      for="switch"
    >
      <span
        class="toggle-switch-inner"
        data-no="Months"
        data-yes="Weeks"
      />
      <span
        class="toggle-switch-switch"
      />
    </label>
  </div>
  <div>
    <label
      class="debounced-input__label"
      for="subreddit"
    >
      Create new team
      <input
        class="debounced-input__input"
        data-testid="debouncedInput"
        id="subreddit"
        value="Team 1"
      />
    </label>
    <button
      class="btn btn-primary m-5"
    >
      Click to create
    </button>
  </div>
  <div
    class="popup__container"
    data-testid="popup__container"
  >
    <div
      class="popup__content bg-light"
    >
      Team wasn't added as it actually exists one team with the same name
    </div>
  </div>
</div>
`;

exports[`Test suite for TeamList should render and be in the document 1`] = `
<div>
  <label
    class="debounced-input__label"
    for="subreddit"
  >
    Scans to fetch
    <input
      class="debounced-input__input"
      data-testid="debouncedInput"
      id="subreddit"
      value=""
    />
  </label>
  <div
    class="toggle-switch m-5"
    data-testid="toggle-switch"
  >
    <input
      class="toggle-switch-checkbox"
      id="switch"
      name="switch"
      readonly=""
      type="checkbox"
    />
    <label
      class="toggle-switch-label"
      for="switch"
    >
      <span
        class="toggle-switch-inner"
        data-no="Months"
        data-yes="Weeks"
      />
      <span
        class="toggle-switch-switch"
      />
    </label>
  </div>
  <div>
    <label
      class="debounced-input__label"
      for="subreddit"
    >
      Create new team
      <input
        class="debounced-input__input"
        data-testid="debouncedInput"
        id="subreddit"
        value=""
      />
    </label>
    <button
      class="btn btn-primary m-5"
    >
      Click to create
    </button>
  </div>
</div>
`;
